server/svix-server/src/v1/utils/mod.rs

```diff
@@ -563,6 +564,95 @@ pub struct EventTypeNamePath {
     pub event_type_name: EventTypeName,
 }
 
+/// JsonStatus is a wrapper over `axum::extract::Json` as a handler output.
+/// Setting the `STATUS` const parameter automatically sets the response
+/// status code, as well as inserting it into the aide documentation.
+pub struct JsonStatus<const STATUS: u16, T: JsonSchema + Serialize>(pub T);
+
+impl<const STATUS: u16, T: JsonSchema + Serialize> IntoResponse for JsonStatus<STATUS, T> {
+    fn into_response(self) -> axum::response::Response {
+        (
+            StatusCode::from_u16(STATUS).unwrap(),
+            axum::extract::Json(self.0),
+        )
+            .into_response()
+    }
+}
+
+impl<const STATUS: u16, T: JsonSchema + Serialize> OperationOutput for JsonStatus<STATUS, T> {
+    type Inner = T;
+
+    fn operation_response(
+        ctx: &mut aide::gen::GenContext,
+        operation: &mut aide::openapi::Operation,
+    ) -> Option<aide::openapi::Response> {
+        axum::extract::Json::<T>::operation_response(ctx, operation)
+    }
+
+    fn inferred_responses(
+        ctx: &mut aide::gen::GenContext,
+        operation: &mut aide::openapi::Operation,
+    ) -> Vec<(Option<u16>, aide::openapi::Response)> {
+        if let Some(resp) = Self::operation_response(ctx, operation) {
+            vec![(Some(STATUS), resp)]
+        } else {
+            vec![]
+        }
+    }
+}
+
+/// JsonStatusUpsert is a wrapper over `axum::extract::Json` as a handler
+/// output. It is similar to `JsonStatus`, the main difference being that
+/// it handles two different status codes for the same response body.
+/// This is a common pattern for handlers that operate in an upsert
+/// fashion.
+pub enum JsonStatusUpsert<const SUPDATE: u16, const SCREATE: u16, T: JsonSchema + Serialize> {
+    Updated(T),
+    Created(T),
+}
+
+impl<const SUPDATE: u16, const SCREATE: u16, T: JsonSchema + Serialize> IntoResponse
+    for JsonStatusUpsert<SUPDATE, SCREATE, T>
+{
+    fn into_response(self) -> axum::response::Response {
+        let (status, body) = match self {
+            JsonStatusUpsert::Updated(v) => (SUPDATE, v),
+            JsonStatusUpsert::Created(v) => (SCREATE, v),
+        };
+        (
+            StatusCode::from_u16(status).unwrap(),
+            axum::extract::Json(body),
+        )
+            .into_response()
+    }
+}
+
+impl<const SUPDATE: u16, const SCREATE: u16, T: JsonSchema + Serialize> OperationOutput
```

Comment:
Looking at this, it looks like we can use the StatusCode enum and we don't need to actually use the ints?

The status code enum has it faults, but I think it's still better than magic numbers everywhere that are easy to forget what they mean exactly everywhere.

Reply:
https://github.com/hyperium/http/blob/master/src/status.rs#L45

`StatusCode` is not an enum, it's a newtype struct, and it has some `const`s defined for its various values (`StatusCode::OK`) which is why it might have looked like an enum.

Optionally, we can hand-roll our own status code enum which we map to `http::StatusCode`.

Reply:
And we can't StatusCode::OK.as_u16()? It's a bit ugly I guess, grr.

OK, let's keep it as is.

Reply:
Ah, no, wait, we don't need the as_u16.

Can't we just accept that, and then inside, in the describe function just do `as_u16`?

Reply:
`as_u16` is not `const fn` so that's not an option (yet). I looked at `http`s issue tracker but no one seems to have even requested it, even though it seems pretty natural.

Reply:
> Ah, no, wait, we don't need the as_u16.
> 
> Can't we just accept that, and then inside, in the describe function just do `as_u16`?

No, const generics can only be integral primitives (iN, uN, bool, etc.) and enums. StatusCode isn't either of those.

Reply:
I mean *inside* of the implementation, but OK, if it has to be a primitive then that won't work.